src/lsm/grid.zig

```diff
@@ -448,22 +509,14 @@ pub fn GridType(comptime Storage: type) type {
             const cache_block = &grid.cache_blocks[cache_index];
             std.mem.swap(BlockPtr, cache_block, completed_write.block);
             std.mem.set(u8, completed_write.block.*, 0);
+            grid.cache_valid.set(cache_index);
```

Comment:
How can we remove the need for `cache_valid` entirely?

My concern is that `cache_valid` adds a dimension of unencapsulated logic, apart from the SetAssociativeCache, which means we have more if/else branches to keep things in sync (or to assert that things are in sync) wherever we work with the SetAssociativeCache.

Instead, we could add a method to invalidate the existing metadata `counts` bits directly in SetAssociativeCache. This also saves memory.

Reply:
> How can we remove the need for cache_valid entirely?

It isn't really "needed" even now, it just enables extra (powerful) validation of cache consistency.
If 65KB per 32GB is too expensive we could gate `cache_valid` behind `constants.verify`. (Imo it isn't too expensive though.)

> My concern is that cache_valid adds a dimension of unencapsulated logic, apart from the SetAssociativeCache, which means we have more if/else branches to keep things in sync (or to assert that things are in sync) wherever we work with the SetAssociativeCache.

The idea of `cache_valid` is that we do want to keep "invalid" blocks in the cache — they _might_ be valid, we just don't know for certain (yet). Known-valid cache entries can be validated (asserted) much more strictly.

Reply:
Maybe we could do something like: add `get_unsafe` to `SetAssociativeCache` which returns entries even if their `count == 0`. But in addition to increasing the API surface area, that means that "invalid" entries will be overwritten even if they are "hot" keys (which would otherwise have a high count and stay cached longer).

So yes, we are paying an extra bit for `cache_valid`, but we are getting better caching in exchange.

Reply:
Agreed that the safer validation is worth the 64 KiB.

I was wondering more if the optimization around caching during state sync is worth this scaffolding. For example, if state sync didn't require the cache in the first place. If it's not also safer/simpler to do that, to verify what's on disk directly.